File: articles/numerical-methods-and-algorithmic-approximation/fortran/numerical_approximation.f90
program numerical_approximation
  implicit none
  real(8) :: x, h, estimate
  x = 1.0d0
  h = 0.01d0
  estimate = (f(x + h) - f(x - h)) / (2.0d0 * h)
  print '(F16.12)', estimate
contains
  real(8) function f(x)
    real(8), intent(in) :: x
    f = sin(x) + 0.25d0 * x * x
  end function f
end program numerical_approximation

File: articles/numerical-methods-and-algorithmic-approximation/prolog/numerical_approximation.pl
f(X, Y) :- Y is sin(X) + 0.25 * X * X.
central_difference(X, H, Estimate) :-
    Xp is X + H,
    Xm is X - H,
    f(Xp, Fp),
    f(Xm, Fm),
    Estimate is (Fp - Fm) / (2 * H).
